sdk/python/lib/pulumi/provider/experimental/provider.py

```diff
@@ -0,0 +1,738 @@
+        :param new_inputs: The new inputs for the resource.
+        :param ignore_changes: A list of properties to ignore when computing the difference.
+        """
+        self.urn = urn
+        self.resource_id = resource_id
+        self.old_state = old_state
+        self.new_inputs = new_inputs
+        self.ignore_changes = ignore_changes
+
+    @property
+    def type(self) -> str:
+        """
+        Extracts the type from the URN.
+        """
+        return _extract_type(self.urn)
+
+    @property
+    def name(self) -> str:
+        """
+        Extracts the name from the URN.
+        """
+        return _extract_name(self.urn)
+
+
+class PropertyDiffKind(Enum):
+    """
+    Represents the kind of difference for a property.
+    """
+
+    ADD = 0
+    ADD_REPLACE = 1
+    DELETE = 2
+    DELETE_REPLACE = 3
+    UPDATE = 4
+    UPDATE_REPLACE = 5
+
+
+class PropertyDiff:
+    """
+    Represents a difference in a property.
+    """
+
+    def __init__(self, kind: PropertyDiffKind, input_diff: bool) -> None:
+        """
+        :param kind: The kind of difference.
+        :param input_diff: Whether the difference is in the input.
+        """
+        self.kind = kind
+        self.input_diff = input_diff
+
+
+class DiffResponse:
+    """
+    Represents the response of a diff operation.
+    """
+
+    def __init__(
+        self,
+        changes: Optional[bool] = None,
+        replaces: Optional[List[str]] = None,
+        stables: Optional[List[str]] = None,
+        delete_before_replace: bool = False,
+        diffs: Optional[List[str]] = None,
+        detailed_diff: Optional[Dict[str, PropertyDiff]] = None,
+    ) -> None:
+        """
+        :param changes: Whether there are changes.
+        :param replaces: List of properties that require replacement.
+        :param stables: List of stable properties.
+        :param delete_before_replace: Whether to delete before replacement.
+        :param diffs: List of properties that have differences.
+        :param detailed_diff: Detailed differences for properties.
+        """
+        self.changes = changes
+        self.replaces = replaces or []
+        self.stables = stables or []
+        self.delete_before_replace = delete_before_replace
+        self.diffs = diffs or []
+        self.detailed_diff = detailed_diff or {}
+
+
+class InvokeRequest:
+    """
+    Represents a request to invoke a provider function.
+    """
+
+    def __init__(self, tok: str, args: Dict[str, PropertyValue]) -> None:
+        """
+        :param tok: The token identifying the function to invoke.
+        :param args: The arguments for the function.
+        """
+        self.tok = tok
+        self.args = args
+
+
+class InvokeResponse:
+    """
+    Represents the response from invoking a provider function.
+    """
+
+    def __init__(
+        self,
+        return_value: Optional[Dict[str, PropertyValue]] = None,
+        failures: Optional[List[CheckFailure]] = None,
```

Comment:
Should `CheckFailure` be renamed to just `Failure` if we're re-using it here? Or does this naming match other existing code?

Reply:
It matches the other libraries, and the protobuf definition. I agree it's not great... might be we should just rename it `Failure` even if we can't rename the proto definitions.
Both Go (https://github.com/pulumi/pulumi-go-provider/blob/f7200e89639601330bba207de374aebab7d958d8/provider.go#L219) and dotnet (https://github.com/pulumi/pulumi-dotnet/blob/main/sdk/Pulumi/Provider/Provider.cs#L214) would need updating though.

Reply:
I am fine with either choice. I do agree with trying to keep it consistent, so if we rename it here, we should also do it in the other languages.